### PR TITLE
Adding an option to specify a target URL

### PR DESCRIPTION
**Pull request by dwelch-r7**
This PR is for an additional option that allows users to specify a URL (i.e. `set RHOST_URL https://example.com/path`) rather than setting each individual option that is the current workflow (i.e. `set RHOSTS example.com`, `set RPORT 443`, `set SSL true` ... etc)

For full details on the various approaches to adding URL Support to Metasploit:
https://github.com/rapid7/metasploit-framework/wiki/RFC---Metasploit-URL-support

I currently only have the new option added to `Exploit::Remote::HttpClient` but I'm sure there are a few other places it also belongs, more than happy to add it in anywhere that makes sense

Here is an example of the setting in action:
image

The reverse is also true, i.e. you can set multiple individual values and the RHOST_URL will be displayed for you:
image

Http username and password is also supported:
image

Some things to note with the current implementation:
- Only single URLs may be specified (In other words you are not able specify an address range in the new option), attempting this will blank out the RHOST_URL option, but any previous options that were set will remain.
image

- URLs of the format `example.com` notably missing the scheme are not supported, you must do `//example.com:<port>` or even better `https://example.com`
image

To sum up, if this feature is to be added users would be able to simply input `set RHOST_URL` and paste in their target rather than having to set anywhere up to 6 different options depending on what that particular module cared about.

### CODE REVIEW

File: lib/msf/core/opt_rhost_url.rb [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+    def calculate_value(datastore)
+      if datastore['RHOSTS']
+        begin
+          uri = URI::Generic.build(host: datastore['RHOSTS'])
+          uri.port=datastore['RPORT']
+          # The datastore uses both `TARGETURI` and `URI` to denote the path of a URL, we try both here and fall back to `/`
+          uri.path=(datastore['TARGETURI'] || datastore['URI'] || '/')
+          uri.user=datastore['HttpUsername']
+          uri.password=datastore['HttpPassword'] if uri.user
```
Comment (adfoster-r7) [inline lib/msf/core/opt_rhost_url.rb:52]: Looks like the protocol's getting dropped at the minute, can we slice in the scheme too? 👀 

```
> set RHOST_URL http://google.com
RHOST_URL => //google.com:80
```

Comment (dwelch-r7) [reply lib/msf/core/opt_rhost_url.rb:52]: Not in a reliable way that I was able to come up with no
see we aren't actually storing the scheme on our end, we'd have to sort of reverse engineer the scheme from the provided options but there's no guarantee we'd get it right, we could end up having the user set a url then immediately be showing them one with a different scheme

Comment (dwelch-r7) [reply lib/msf/core/opt_rhost_url.rb:52]: I think what I could do though is slice off the leading `//` that would solve the problem I think

Comment (adfoster-r7) [reply lib/msf/core/opt_rhost_url.rb:52]: It's an interesting discussion point. The original [RFC](https://github.com/rapid7/metasploit-framework/wiki/RFC---Metasploit-URL-support#additional-considerations) made notes about different schemes too - such as `ftp`, `postgres`, `smb`, etc.

Comment (dwelch-r7) [reply lib/msf/core/opt_rhost_url.rb:52]: Unless you think it'll cause more confusion not having those formats work from the get go I still think this current iteration would benefit a majority of users and we can always add mroe use cases down the line when it becomes apparent they are warranted

Comment (adfoster-r7) [reply lib/msf/core/opt_rhost_url.rb:52]: I'm just keeping an open mind for now, discussing how the different implementations could impact various workflows. We want to reduce the possibility of confusing users in the initial release, and avoiding potential confusion for when we do support more use cases in the future. Finding the right balance between usable, shippable, and iterable will be good :+1:

**Comment by adfoster-r7** [comment]
@dwelch-r7 Are there any additional edge cases to consider? For instance not having a module loaded, setting RHOST_URL, then loading the module and running it? Would that cause any bugs?

```
setg RHOST_URL http://example.com/foo
use scanner/http/title
run
```

Or running a module without setting any url information:

```
use scanner/http/title
run
```

Which I think will cause a crash with the current implementation

**Comment by dwelch-r7** [comment]
> @dwelch-r7 Are there any additional edge cases to consider? For instance not having a module loaded, setting RHOST_URL, then loading the module and running it? Would that cause any bugs or unexpected weirdness?
> 
> ```
> setg RHOST_URL http://example.com/foo
> use scanner/http/title
> run
> ```
> 
> Or running a module without setting any url information:
> 
> ```
> use scanner/http/title
> run
> ```
> 
> Which I think will cause a crash with the current implementation

uhhh yea looks like you're right, I did not realise it was trying to validate it on setting _and_ when running the module, I'll figure something out

**Comment by adfoster-r7** [comment]
Spotted another edgecase:

```
msf5 exploit(windows/http/rejetto_hfs_exec) > set RHOST_URL http://10.10.87.250:8080
RHOST_URL => 10.10.87.250:8080
msf5 exploit(windows/http/rejetto_hfs_exec) > options

Module options (exploit/windows/http/rejetto_hfs_exec):

   Name       Current Setting    Required  Description
   ----       ---------------    --------  -----------
   HTTPDELAY  10                 no        Seconds to wait before terminating web server
   Proxies                       no        A proxy chain of format type:host:port[,type:host:port][...]
   RHOSTS     10.10.87.250       yes       The target host(s), range CIDR identifier, or hosts file with syntax 'file:<path>'
   RHOST_URL  10.10.87.250:8080  no        The target URL, only applicable if there is a single URL
   RPORT      8080               yes       The target port (TCP)
   SRVHOST    0.0.0.0            yes       The local host or network interface ...
   SRVPORT    8080               yes       The local port to listen on.
   SSL        false              no        Negotiate SSL/TLS for outgoing connections
   SSLCert                       no        Path to a custom SSL certificate (default is randomly generated)
   TARGETURI                     yes       The path of the web application
   URIPATH                       no        The URI to use for this exploit (default is random)
   VHOST                         no        HTTP server virtual host

msf5 exploit(windows/http/rejetto_hfs_exec) > run

[-] Exploit failed: bad URI(is not URI?): "10.10.87.250:8080"
[*] Exploit completed, but no session was created.
```

Looks like it doesn't work well for modules that require a `TARGETURI`, the default value was `/`

**Comment by dwelch-r7** [comment]
> Spotted another edgecase:
> 
> ```
> msf5 exploit(windows/http/rejetto_hfs_exec) > set RHOST_URL http://10.10.87.250:8080
> RHOST_URL => 10.10.87.250:8080
> msf5 exploit(windows/http/rejetto_hfs_exec) > options
> 
> Module options (exploit/windows/http/rejetto_hfs_exec):
> 
>    Name       Current Setting    Required  Description
>    ----       ---------------    --------  -----------
>    HTTPDELAY  10                 no        Seconds to wait before terminating web server
>    Proxies                       no        A proxy chain of format type:host:port[,type:host:port][...]
>    RHOSTS     10.10.87.250       yes       The target host(s), range CIDR identifier, or hosts file with syntax 'file:<path>'
>    RHOST_URL  10.10.87.250:8080  no        The target URL, only applicable if there is a single URL
>    RPORT      8080               yes       The target port (TCP)
>    SRVHOST    0.0.0.0            yes       The local host or network interface ...
>    SRVPORT    8080               yes       The local port to listen on.
>    SSL        false              no        Negotiate SSL/TLS for outgoing connections
>    SSLCert                       no        Path to a custom SSL certificate (default is randomly generated)
>    TARGETURI                     yes       The path of the web application
>    URIPATH                       no        The URI to use for this exploit (default is random)
>    VHOST                         no        HTTP server virtual host
> 
> msf5 exploit(windows/http/rejetto_hfs_exec) > run
> 
> [-] Exploit failed: bad URI(is not URI?): "10.10.87.250:8080"
> [*] Exploit completed, but no session was created.
> ```
> 
> Looks like it doesn't work well for modules that require a `TARGETURI`, the default value was `/`

Is that not what you'd expect? I mean if you set an empty path/TARGETURI that's what you should get no?

**Comment by dwelch-r7** [comment]
> Spotted another edgecase:
> 
> ```
> msf5 exploit(windows/http/rejetto_hfs_exec) > set RHOST_URL http://10.10.87.250:8080
> RHOST_URL => 10.10.87.250:8080
> msf5 exploit(windows/http/rejetto_hfs_exec) > options
> 
> Module options (exploit/windows/http/rejetto_hfs_exec):
> 
>    Name       Current Setting    Required  Description
>    ----       ---------------    --------  -----------
>    HTTPDELAY  10                 no        Seconds to wait before terminating web server
>    Proxies                       no        A proxy chain of format type:host:port[,type:host:port][...]
>    RHOSTS     10.10.87.250       yes       The target host(s), range CIDR identifier, or hosts file with syntax 'file:<path>'
>    RHOST_URL  10.10.87.250:8080  no        The target URL, only applicable if there is a single URL
>    RPORT      8080               yes       The target port (TCP)
>    SRVHOST    0.0.0.0            yes       The local host or network interface ...
>    SRVPORT    8080               yes       The local port to listen on.
>    SSL        false              no        Negotiate SSL/TLS for outgoing connections
>    SSLCert                       no        Path to a custom SSL certificate (default is randomly generated)
>    TARGETURI                     yes       The path of the web application
>    URIPATH                       no        The URI to use for this exploit (default is random)
>    VHOST                         no        HTTP server virtual host
> 
> msf5 exploit(windows/http/rejetto_hfs_exec) > run
> 
> [-] Exploit failed: bad URI(is not URI?): "10.10.87.250:8080"
> [*] Exploit completed, but no session was created.
> ```
> 
> Looks like it doesn't work well for modules that require a `TARGETURI`, the default value was `/`

Pushed up a fix for that, I had that in there at one point and removed it for some reason

**Comment by adfoster-r7** [comment]
@dwelch-r7 I think I was expecting semantics similar to curl which handles that scenario gracefully:

```
$ curl -v http://example.com
...
> GET / HTTP/1.1
...
```

File: lib/msf/core/opt_rhost_url.rb (hunk 1 above)
Comment (adfoster-r7) [reply lib/msf/core/opt_rhost_url.rb:53]: I wonder if things become easier if this was renamed from the current generic name to `OptHttpUrl` ? 🤔

**Comment by adfoster-r7** [comment]
We should double check the ability to check this works with `run OPTIONS` on the command line

File: lib/msf/core/opt_rhost_url.rb (hunk 1 above)
Comment (wvu-r7) [reply lib/msf/core/opt_rhost_url.rb:55]: Alan fights for the users.

File: lib/msf/ui/console/command_dispatcher/core.rb [hunk 2]
```diff
@@ -1602,6 +1590,22 @@ def cmd_set(*args)
...
 
+    set_option(name, value, global: global, append: append)
+  end
+
+  #
+  # Sets a name to a value in a context aware environment.
+  #
+  def set_option(name, value, global: false, append: false)
+
+    # Determine which data store we're operating on
```
Comment (wvu-r7) [inline lib/msf/ui/console/command_dispatcher/core.rb:1601]:
```suggestion
    # Determine which datastore we're operating on
```

File: lib/msf/ui/console/command_dispatcher/core.rb [hunk 3]
```diff
@@ -1602,6 +1590,22 @@ def cmd_set(*args)
...
+    if active_module and !global
+      datastore = active_module.datastore
+    else
+      global = true
+      datastore = self.framework.datastore
+    end
```
Comment (wvu-r7) [inline lib/msf/ui/console/command_dispatcher/core.rb:1607]:
```suggestion
  # Sets a name to a value in a context-aware environment.
  #
  def set_option(name, value, global: false, append: false)
    # Determine which datastore we're operating on
    if active_module && !global
      datastore = active_module.datastore
    else
      global = true
      datastore = framework.datastore
    end
```

File: lib/msf/ui/console/command_dispatcher/core.rb [hunk 4]
```diff
@@ -1542,53 +1542,41 @@ def cmd_set_help
...
-          (global) ? "Global" : "Module: #{active_module.refname}",
-          datastore) + "\n")
+                Msf::Serializer::ReadableText.dump_datastore(
+                    (global) ? "Global" : "Module: #{active_module.refname}",
+                    datastore) + "\n")
       return true
-    elsif (args.length == 1)
-      if (not datastore[args[0]].nil?)
+    elsif args.length == 1
+      if not datastore[args[0]].nil?
```
Comment (wvu-r7) [inline lib/msf/ui/console/command_dispatcher/core.rb:1579]:
```suggestion
      if datastore[args[0]]
```

File: lib/msf/ui/console/command_dispatcher/core.rb [hunk 5]
```diff
@@ -1542,53 +1542,41 @@ def cmd_set_help
...
       end
 
       # Dump the active datastore
       print("\n" +
-        Msf::Serializer::ReadableText.dump_datastore(
-          (global) ? "Global" : "Module: #{active_module.refname}",
-          datastore) + "\n")
+                Msf::Serializer::ReadableText.dump_datastore(
+                    (global) ? "Global" : "Module: #{active_module.refname}",
+                    datastore) + "\n")
```
Comment (wvu-r7) [inline lib/msf/ui/console/command_dispatcher/core.rb:1576]: I can't tell what this formatting change is for.

Comment (wvu-r7) [inline lib/msf/ui/console/command_dispatcher/core.rb:1576]: I can't tell what this formatting change is for.

File: lib/msf/ui/console/command_dispatcher/core.rb [hunk 6]
```diff
@@ -1542,53 +1542,41 @@ def cmd_set_help
...
+    if args.length == 0
       # If we aren't dumping the global data store, then go ahead and
       # dump it first
-      if (!global)
+      unless global
         print("\n" +
-          Msf::Serializer::ReadableText.dump_datastore(
-            "Global", framework.datastore))
+                  Msf::Serializer::ReadableText.dump_datastore(
+                      "Global", framework.datastore))
```
Comment (wvu-r7) [inline lib/msf/ui/console/command_dispatcher/core.rb:1569]: I can't tell what this formatting change is for.

File: lib/msf/core/option_container.rb [hunk 7]
```diff
@@ -18,6 +18,7 @@ module Msf
   autoload :OptRaw, 'msf/core/opt_raw'
   autoload :OptRegexp, 'msf/core/opt_regexp'
   autoload :OptString, 'msf/core/opt_string'
+  autoload :OptRhostUrl, 'msf/core/opt_rhost_url'
```
Comment (wvu-r7) [inline lib/msf/core/option_container.rb:21]:
```suggestion
  autoload :OptRhostURL, 'msf/core/opt_rhost_url'
```

https://github.com/rubocop-hq/ruby-style-guide#camelcase-for-classes

File: lib/msf/core/data_store.rb [hunk 8]
```diff
@@ -43,14 +43,23 @@ def []=(k, v)
       end
     end
 
-    super(k,v)
+    if v.is_a? Hash
+      v.each { |key, value| self[key] = value }
+    else
+      super(k,v)
```
Comment (wvu-r7) [inline lib/msf/core/data_store.rb:49]:
```suggestion
      super(k, v)
```

File: lib/msf/core/data_store.rb [hunk 9]
```diff
@@ -43,14 +43,23 @@ def []=(k, v)
...
+    end
   end
 
   #
   # Case-insensitive wrapper around hash lookup
   #
   def [](k)
-    super(find_key_case(k))
+    k = find_key_case(k)
```
Comment (wvu-r7) [inline lib/msf/core/data_store.rb:57]:
```suggestion
    k = find_key_case(k)
    
```

File: lib/msf/core/opt.rb [hunk 10]
```diff
@@ -62,6 +62,10 @@ def self.SSLVersion
       )
     end
 
+    def self.RHOST_URL(default=nil, required=false, desc="The target URL, only applicable if there is a single URL")
```
Comment (wvu-r7) [inline lib/msf/core/opt.rb:65]:
```suggestion
    def self.RHOST_URL(default = nil, required = false, desc = 'The target URL, only applicable if there is a single URL')
```

https://github.com/rubocop-hq/ruby-style-guide#spaces-around-equals

File: lib/msf/core/opt.rb [hunk 11]
```diff
@@ -62,6 +62,10 @@ def self.SSLVersion
...
+      Msf::OptRhostUrl.new(__method__.to_s, [ required, desc, default ])
```
Comment (wvu-r7) [inline lib/msf/core/opt.rb:66]: https://github.com/rapid7/metasploit-framework/pull/13766/files#r451874140

File: lib/msf/core/opt_rhost_url.rb [hunk 12]
```diff
@@ -0,0 +1,75 @@
+# -*- coding: binary -*-
+
+module Msf
+  ###
+  #
+  # RHOST URL option.
+  #
+  ###
+  class OptRhostUrl < OptBase
```
Comment (wvu-r7) [inline lib/msf/core/opt_rhost_url.rb:9]: https://github.com/rapid7/metasploit-framework/pull/13766/files#r451874140

File: lib/msf/core/opt_rhost_url.rb [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+    def type
+      'rhost url'
+    end
+
+
```
Comment (wvu-r7) [inline lib/msf/core/opt_rhost_url.rb:14]:
```suggestion
```

File: lib/msf/core/opt_rhost_url.rb [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+      return unless value
+
+      uri = get_uri(value)
+      option_hash = {}
+      # Blank this out since we don't know if this new value will have a `VHOST` to ensure we remove the old value
+      option_hash['VHOST'] = nil
+
+      option_hash['RHOSTS'] = uri.hostname
+      option_hash['RPORT'] = uri.port
+      option_hash['SSL'] = %w{ssl https}.include?(uri.scheme)
```
Comment (wvu-r7) [inline lib/msf/core/opt_rhost_url.rb:25]:
```suggestion
      option_hash['SSL'] = %w[ssl https].include?(uri.scheme)
```

https://github.com/rubocop-hq/ruby-style-guide#percent-literal-braces

Comment (adfoster-r7) [reply lib/msf/core/opt_rhost_url.rb:25]: During module hacking we discussed just running Rubocop over this file in its entirety, there's a things to lint such as a few unneeded uses of `and` etc.

File: lib/msf/core/opt_rhost_url.rb [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+
+      # Both `TARGETURI` and `URI` are used as datastore options to denote the path on a uri
+      option_hash['TARGETURI'] = uri.path || '/'
+      option_hash['URI'] = uri.path || '/'
+
+      if uri.scheme and %{http https}.include?(uri.scheme)
```
Comment (wvu-r7) [inline lib/msf/core/opt_rhost_url.rb:31]: https://github.com/rapid7/metasploit-framework/pull/13766/files#r451880965

Comment (wvu-r7) [inline lib/msf/core/opt_rhost_url.rb:31]:
```suggestion
      if uri.scheme && %[http https].include?(uri.scheme)
```

https://github.com/rapid7/metasploit-framework/pull/13766/files#r451880965

File: lib/msf/core/opt_rhost_url.rb [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+        option_hash['HttpPassword'] = uri.password.to_s
+      end
+
+      option_hash
+    end
+
+    def valid?(value, check_empty: false)
+      return true unless value
+      uri = get_uri(value)
+      false unless uri.host != nil and uri.port != nil
```
Comment (wvu-r7) [inline lib/msf/core/opt_rhost_url.rb:43]:
```suggestion
      return false if uri.host.nil? && uri.port.nil?
```

Did you mean to `return`? What happens with the `super`?

File: lib/msf/core/opt_rhost_url.rb [hunk 17]
```diff
@@ -0,0 +1,75 @@
...
+    protected
+
+    def get_uri(value)
+      begin
+        uri = URI(value)
+      rescue URI::InvalidURIError
+        uri = URI('//' + value)
+      end
+      uri
+    end
```
Comment (wvu-r7) [inline lib/msf/core/opt_rhost_url.rb:72]:
```suggestion
    def get_uri(value)
      URI(value)
    rescue URI::InvalidURIError
      URI("//#{value}")
    end
```

https://github.com/rubocop-hq/ruby-style-guide#implicit-begin

File: lib/msf/core/opt_rhost_url.rb [hunk 18]
```diff
@@ -0,0 +1,75 @@
...
+
+    def valid?(value, check_empty: false)
+      return true unless value
+      uri = get_uri(value)
+      false unless uri.host != nil and uri.port != nil
+      super
+    end
+
+    def calculate_value(datastore)
+      if datastore['RHOSTS']
```
Comment (wvu-r7) [inline lib/msf/core/opt_rhost_url.rb:48]: Guard clause, please?

https://github.com/rubocop-hq/ruby-style-guide#nested-conditionals

File: lib/msf/core/opt_rhost_url.rb [hunk 19]
```diff
@@ -0,0 +1,75 @@
...
+        begin
```
Comment (wvu-r7) [inline lib/msf/core/opt_rhost_url.rb:49]: https://github.com/rapid7/metasploit-framework/pull/13766/files#r451890300

**Review by wvu — CHANGES_REQUESTED** [review]
It might be a good idea to use RuboCop.

File: lib/msf/core/opt_rhost_url.rb (hunk 14 above)
Comment (wvu-r7) [reply lib/msf/core/opt_rhost_url.rb]: https://github.com/rapid7/metasploit-framework/pull/13766#pullrequestreview-445207149

Sounds good.

File: lib/msf/ui/console/command_dispatcher/core.rb [hunk 20]
```diff
@@ -1602,6 +1590,22 @@ def cmd_set(*args)
     name  = args[0]
     value = args[1, args.length-1].join(' ')
 
+    set_option(name, value, global: global, append: append)
+  end
+
+  #
+  # Sets a name to a value in a context aware environment.
+  #
+  def set_option(name, value, global: false, append: false)
```
Comment (adfoster-r7) [inline lib/msf/ui/console/command_dispatcher/core.rb:1599]: Can we revert this file now? 🤔

File: lib/msf/core/exploit/http/client.rb [hunk 21]
```diff
@@ -22,6 +22,7 @@ def initialize(info = {})
 
     register_options(
```
Comment (adfoster-r7) [inline lib/msf/core/exploit/http/client.rb:23]:
```
if framework.features.enabled?("feature_name")
  register_options([OPT::RHOST_URL])
end
```

Comment (adfoster-r7) [reply lib/msf/core/exploit/http/client.rb:23]: Would you mind pulling the latest of the https://github.com/rapid7/metasploit-framework/pull/13833 PR to get the latest features API

File: lib/msf/ui/console/command_dispatcher/core.rb (hunk 20 above)
Comment (adfoster-r7) [reply lib/msf/ui/console/command_dispatcher/core.rb:1777]: @dwelch-r7 Just confirming if this file can be reverted or not :+1:

Comment (dwelch-r7) [reply lib/msf/ui/console/command_dispatcher/core.rb:1777]: It's not needed for these changes no, but I think they're good changes, I can revert and put it up as a separate PR, that probably makes more sense

**Pull request closed without merging**